### DISCUSSION TITLE
Proposal: introduce left and right join clauses to C# query expression syntax

### DISCUSSION THREAD

**Discussion by roji** [dotnet/csharplang#8892] [Language Ideas]
# Summary

- [x] Specification: https://github.com/dotnet/csharplang/blob/main/proposals/left-right-join-in-query-expressions.md
- [ ] Tracking issue: https://github.com/dotnet/csharplang/issues/8947

Introduce `left join` and `right join` clauses into LINQ query expression syntax, e.g.:

```c#
from student in students
left join department in departments on student.DepartmentID equals department.ID
select new { student.Name, department?.Name }
```

# Motivation

LINQ has a [Join operator](https://learn.microsoft.com/dotnet/api/system.linq.enumerable.join), which, like its SQL INNER JOIN counterpart, correlates elements of two sequences based on matching keys; the C# language has a corresponding `join` clause which translates to this operator ([section 12.20.3.5 of the C# specs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#122035-from-let-where-join-and-orderby-clauses)).

In addition to INNER JOIN, SQL also has LEFT JOIN, which returns outer elements even if there's no corresponding inner ones; LINQ and C#, in contrast, lack this operator. [The LINQ conceptual documentation shows how to combine existing operators to achieve a left join](https://learn.microsoft.com/en-us/dotnet/csharp/linq/standard-query-operators/join-operations#perform-left-outer-joins):

```c#
var query =
    from student in students
    join department in departments on student.DepartmentID equals department.ID into gj
    from subgroup in gj.DefaultIfEmpty()
    select new
    {
        student.FirstName,
        student.LastName,
        Department = subgroup?.Name ?? string.Empty
    };
```

Or using method syntax:

```c#
var query = students
    .GroupJoin(departments, student => student.DepartmentID, department => department.ID, (student, departmentList) => new { student, subgroup = departmentList })
    .SelectMany(
        joinedSet => joinedSet.subgroup.DefaultIfEmpty(),
        (student, department) => new
        {
            student.student.FirstName,
            student.student.LastName,
            Department = department.Name
        });
```

Although functionality sufficient for expressing a left join operation, this combining approach has the following drawbacks:

- It's complicated, requiring combining multiple different LINQ operators in a specific way to form a complex construct, and is easy to accidentally get wrong. Many EF users have complained about the complexity of this construct for expressing a simple SQL LEFT JOIN.
- It is inefficient - the combination of operators adds significant overhead compared to a single operator using an internal lookup table (or "hash join", as Join is implemented).

https://github.com/dotnet/runtime/issues/110292 is an approved API proposal for introducing LeftJoin and RightJoin operators into the System.Linq for .NET 10; see that proposal for additional performance information and benchmarks. Aside from allowing users to more easily express SQL LEFT/RIGHT JOIN when using a LINQ provider (such as EF Core), it would also allow both easier and more efficient in-memory left/right joins, exactly in the way that inner joins are already supported.

Along with the introduction of the operators into System.Linq, the C# query expression syntax can be extended with new `left join` and `right join` clauses which would translate to these new operators, allowing for simpler and more efficient code where C# query syntax is used.

# Design

The syntax for the proposed `left join` and `right join` expression clauses would be identical to the existing `join` clause. For example:

```c#
from student in students
left join department in departments on student.DepartmentID equals department.ID
select new { student.Name, department?.Name }
```

Note that after the `left join` clause, `department` is nullable, as `left join` also returns students with no corresponding departments (unlike the existing `join` clause, which does not).

Similar to how the `join` clause internally translates to the LINQ `Join` operator, `left join` would translate to the new `LeftJoin` operator being added in https://github.com/dotnet/runtime/issues/110292:

```c#
students.LeftJoin(
   departments,
   student => student.DepartmentID,
   department => department.CustomerID,
   (student, department) => new { student.Name, department?.Name })
```

`right join` represents the opposite operation, returning inner items when no corresponding outer ones are found:

```c#
from department in departments
right join student in students on department.ID equals student.DepartmentID
select new { department?.Name, student.Name }
```

This would translate to:

```c#
departments.RightJoin(
   students,
   department => department.CustomerID,
   student => student.DepartmentID,
   (department, student) => new { department?.Name, student.Name })
```

# Notes

* See https://github.com/dotnet/runtime/issues/110292 for discussion on the naming for the new operators, and motivations for the adopted LeftJoin and RightJoin naming. For the proposed C# query expression, the clauses could be named `left join`/`right join`, or possibly `leftjoin`/`rightjoin` (similar to the `orderby` clause).

/cc @eiriktsarpalis @terrajobst @CyrusNajmabadi @dotnet/efteam

### Previous related issues

* https://github.com/dotnet/csharplang/discussions/361
* https://github.com/dotnet/csharplang/issues/508
* https://github.com/dotnet/runtime/issues/53236
* https://github.com/dotnet/efcore/issues/12793

**Comment by TahirAhmadov** [comment]
Minor typo, second example should have right join, not left join.

**Reply by roji** [reply]
Thanks, corrected.

**Comment by houbi56** [comment]
Please also include a FullJoin!

**Reply by roji** [reply]
This is out of scope for this proposal. FullJoin has not been asked for much, and an efficient implementation requires more careful thinking compared to Left/RightJoin. If user demand for this manifests, we can always introduce this in the future.

**Comment by fubar-coder** [comment]
@houbi56 The API for left/right outer join was already approved during the [.NET API review on 2024-12-03](https://youtu.be/itbo6fSB9nA?t=645), but full join was omitted, because no demand was seen for this operation.

**Comment by HaloFour** [comment]
Kinda surprised this made it through review so easily.  My opinion has been that new methods are just as problematic as new expression nodes, in that existing queryable providers must be updated to support both or they will explode at runtime.  Obviously that's true of any and all methods, but such fundamental operations as `LeftJoin` and `RightJoin` might create an expectation of more universal support from developers, who now have to know whether they can use these methods/clauses or fall back to using the old GroupJoin approach depending on what they are targeting.

That's not to say that I'm opposed to the addition of these new methods, quite the contrary.  I hope that their addition might help to open up the wider conversation about adding new expression nodes so that expression trees can support newer language features as well.

**Reply by KennethHoff** [reply]
I agree. The double standard of "Can't fix expression trees because we would break pre-existing LINQ providers" and "We can add new LINQ methods because LINQ providers will adapt" is so weird. 

Either just say "No changes to LINQ" or fix expression trees. This is literally the worst where you can't trust them to work while also requiring you to use legacy language features.

**Reply by CyrusNajmabadi** [reply]
This is the same as any other method afaict wrt to adding new methods to the BCL.   I don't remember any methods not being added because providers would not understand them without updating.

**Reply by HaloFour** [reply]
> This is the same as any other method afaict wrt to adding new methods to the BCL.

I disagree, only because these particular methods are intended to be used in very common LINQ scenarios, and the risk of them being used with queryable providers that don't understand them is high as a result.

Otherwise I do agree, adding new methods and using them in LINQ queries already breaks queryable providers.  There doesn't seem to be a huge concern about that.  So it doesn't make a lot of sense to me that there is such a huge concern when it comes to adding expression nodes and allowing LINQ expressions to catch up with the language.

**Reply by CyrusNajmabadi** [reply]
> I disagree, only because these particular methods are intended to be used in very common LINQ scenarios

Linq is widely used outside of expression trees. It has had active involvement and improvement since inception.  It's never been locked down. Only expr trees had that happen.  And that was not popular. I'm not sure why we'd take the policy and apply to areas that haven't had issues and are still popular and in wide use.

**Reply by jnm2** [reply]
If "providers won't understand new node type" holds us back from letting expressions use `?.`, why doesn't "providers won't understand new LINQ methods tied to new language syntax" also carry the same weight?

**Reply by CyrusNajmabadi** [reply]
> and the risk of them being used with queryable providers that don't understand them is high as a result.

No new nodes are being added here afaict.  The compilation of these methods is the same as any other methods one could/would use in an expression tree. I don't know about any rule or desire to not add normal in-memory operations because someone would not be able to to use them in an queryable context. 

Truly, I genuinely don't remember any situation where such a decision was made in the past. If the linq owners do have that as a rule, then fine. But I would be surprised this made it through API review if that were the case.

**Reply by CyrusNajmabadi** [reply]
> providers won't understand new LINQ methods tied to new language syntax

What new language syntax?  I don't think anyone has proposed or is championing that. 

Afaict, this is just a new method in the bcl. Something that doesn't have restrictions in this space.

**Reply by HaloFour** [reply]
Indeed, no new nodes are being added here.  But the end result is the same, use of these common query operators will fail with queryable providers.

I'm not arguing against adding these methods, or supporting them through new LINQ query clauses\*, but I don't think that this situation is appreciably different from adding new nodes for the purpose of supporting new language syntax.  Either way, every existing queryable provider will go boom unless they are updated to support this.

\* Quite the opposite, I would love to see robust evolution in this space.

**Reply by CyrusNajmabadi** [reply]
> but I don't think that this situation is appreciably different from adding new nodes for the purpose of supporting new language syntax.

I think it's very different. With this approach, the runtime simply could not as methods at all, lest someone try to use in a queryable.

**Reply by CyrusNajmabadi** [reply]
Whoops. Sorry, I thought this was the runtime discussion. I didn't realize this was csharplang. I don't see us making new lang syntax here unless there is super strong desire on that from a partner or domain.

**Reply by HaloFour** [reply]
> Whoops. Sorry, I thought this was the runtime discussion. I didn't realize this was csharplang. I don't see us making new lang syntax here unless there is super strong desire on that from a partner or domain.

That's fair, but unfortunate.  These operators are common in LINQ queries, and I expect EF will support them, so it'll be weird to not be able to take advantage of them from the language directly unless they use method syntax.  Developers will have to know to use the existing GroupJoin approach, which is awkward.

**Reply by jnm2** [reply]
This thread is proposing adding `left join` and `right join` to LINQ syntax. This would not create new node types, but I don't know why "new `?.` node type that providers die on" is more of a hurdle than "new keywords enshrining methods that providers die on."

**Reply by CyrusNajmabadi** [reply]
New nodes, vs reuse of existing nodes, seems entirely different to me. Especially as it's always been the case you could reference other methods, and we've never blocked that (outside of things like ref structs which can't be represented in an expression tree).

**Reply by CyrusNajmabadi** [reply]
Basically, I have no idea why we'd add restrictions when the Linq owners themselves are fine adding the method. Bcl and Linq have never had a problem adding new method. Why would we disallow it?  Who is asking for it to be blocked?

**Reply by HaloFour** [reply]
Nobody is asking for restrictions, but I would hope the conversation around adding such fundamental and common LINQ methods would open the conversation to supporting them in C# language (as the alternative is kinda hoaky), which might open the door to discussing expression tree evolution in general.  :)

**Reply by CyrusNajmabadi** [reply]
> but I don't know why "new ?. node type that providers die on" 

That's a decision for the expression tree owners. As long as that is their position, we'll follow it. But that decision doesn't extend any further beyond that API afaik.  (Maybe there is a decision in the runtime for the past couple of decades where that differs?). 

But as far as I can remember, Linq was absolutely fair game to enhance.  It's only been the expr trees API that got locked.

**Reply by CyrusNajmabadi** [reply]
> which might open the door to discussing expression tree evolution in general. 

The expr trees owners can definitely decide if they want to change the rules. But it's orthogonal to any bcl API decision to me :-)

**Reply by roji** [reply]
Just to mention that there have been many LINQ method additions in recent years. For example, [the Order operator](https://learn.microsoft.com/en-us/dotnet/api/system.linq.queryable.order) was added in .NET 7 - there's no difference between that and LeftJoin/RightJoin. I absolutely don't think there is - or should be - any decision to not introduce new LINQ operator methods just because they'd create issues with existing queryable LINQ providers.

I personally happen to agree that there's no fundamental difference in terms of compat between adding a new LINQ method and adding a new expression tree node - but I'd really like to keep the conversation around evolving LINQ expression trees out of this language proposal (in the interests of keeping things focused and not making the discussion here unnecessarily complex/confused).

**Comment by CyrusNajmabadi** [comment]
I will champion this.

**Comment by orthoxerox** [comment]
@roji I've watched the API review and I've heard that you don't exclude the possibility of adding anti joins at a later date. Anti joins are a surprisingly useful concept that I use a lot when working with data frames. I know it's just a `LeftJoin` plus `Where`, but a dedicated anti join can be much faster, especially when the left and the right datasets are almost equal.

**Reply by roji** [reply]
We'd need a good perf investigation showing that dedicated SemiJoin/AntiJoin operations consistently out-perform the alternatives - that's definitely a possibility but is out of scope for this language proposal. Note that LeftJoin aren't only about performance, they're also about providing a much simpler alternative to the existing way of expressing LeftJoin (GroupJoin/SelectMany/DefaultIfEmpty); SemiJoin/AntiJoin don't really have this aspects (since they can already be expressed relatively cleanly via existing operators), and also do not have corresponding SQL operations (in SQL, semi-joins are expressed via EXISTS and a subquery, anti-joins via NOT EXISTS).

We can always consider introducing dedicated operators in the future based on the performance improvements and user interest.

**Reply by terrajobst** [reply]
I don't think I've ever seen a SQL dialect that had explicit support for semi joins. It seems to be an implementation detail of the plan generator. The canonical way in SQL would be an EXISTS predicate, which would be `.Any()`, wouldn't it?

**Reply by roji** [reply]
@terrajobst yeah, absolutely - explicit semi/anti join operators aren't really useful for the purposes of translating to SQL. However, the in-memory (enumerable, not queryable) operators could bring significant perf advantages, since they could e.g. use a hash/lookup internally (just like Join and LeftJoin do), which would likely perform better for most scenarios than the alternative with Any (which does a nested loop).

So this really would be a perf feature for in-memory LINQ operations, not for SQL translation.

**Reply by terrajobst** [reply]
Ah, gotcha. Same reason as for outer joins then.

**Reply by roji** [reply]
Yeah, though the outer joins also have the additional reason of being more complicated/annoying to express (with GroupJoin+SelectMany+DefaultIfEmpty) - so it's a usability issue there as well, in addition to the in-memory runtime perf improvement. In contrast, semi/anti don't really have a usability problem today (the Any way of expressing them is pretty good).

**Comment by charlesroddie** [comment]
I am opposed to adding these operators:

1. These are highly complex operations in SQL whose complexity is hidden in simple syntax. The meaning of `SELECT x.A, y.B from TX x LEFT JOIN TY y ON condition` is:
    1. Take the inner join
    2. Take the set of x such that there is no y matching the condition, and map each x in this set to a result row consisting of taking the syntactic structure [x.A, y.B] and for each thing not starting in `y` calculating it and for each subexpression starting in `y` not calculating it but instead replacing it with null.
    3. Take the disjoint union of these two sets.
2. The complexity is not appreciated by most developers. For example almost all SQL tutorials attempt to explain left join with some venn diagram attempting to explain this as "TX" - as "TX" sometimes with augmented data from "Y". This explanation is not valid as it can only work if there is at most 1 matching Y for each X. So even the almost all SQL guide writers are unable to understand or explain these operations. See https://minimalmodeling.substack.com/p/many-explanations-of-join-are-wrong .
3. Most developers using LEFT/RIGHT JOIN use these as normal, simple operations, when in fact the complexity requires that there should be extensive documentation and explanation accompanying each usage to show that it does what is intended. I personally have spent hours fixing bugs from usage of these operators when the original developer did not realize how difficult these operations really are to understand and therefore to write correctly. They are now banned in our repo,

Overall these operators are not a normal part of relational algebra and should never have been in SQL, and their use should be discouraged rather than further enabled. It's right that a fundamentally complex operation should have complex syntax.

**Reply by CyrusNajmabadi** [reply]
@charlesroddie the operators were already added to the runtime.

**Reply by roji** [reply]
> Overall these operators are not a normal part of relational algebra and should never have been in SQL, and their use should be discouraged rather than further enabled. It's right that a fundamentally complex operation should have complex syntax.

This is the first time I hear someone trying to claim that LEFT/RIGHT JOIN shouldn't have been in SQL; and your principle here I really don't get - so many other things (in so many languages, not just SQL) are precisely about hiding complexity behind a simple syntax. Shouldn't GROUP BY be removed because of this as well, or do you somehow not consider it "complex"? What about INNER JOIN? What are the formal criteria for "complex" here in the first place?

In any case, this is no place for a discussion on SQL - at least from the SQL perspective, all we're doing here in .NET is surfacing existing concepts for easier use in .NET, regardless of what anyone's opinion on them is.

**Reply by CyrusNajmabadi** [reply]
Also worth noting, c#'s joins are equijoins, not generalized joins.  So not sure the complaints raised apply here at all.

**Reply by terrajobst** [reply]
SQL has many problems with respect to how it surfaces relational algebra. But the reality is it's still the most widely database language. Surfacing those concepts in Linq (and by extension) C# still make a lot of sense to me.

**Comment by adamburgess** [comment]
@CyrusNajmabadi do you have any updates on this? I am hoping it is still planned for C# 14 as was mentioned in the LDM notes 🙏

**Reply by roji** [reply]
The C# LINQ syntax support for this feature hasn't yet been implemented - when it does is TBD.

**Reply by reichlew** [reply]
I believe the C# LINQ syntax support for this feature has been fully implemented now? Is there any further follow-up on this item?

**Reply by roji** [reply]
This has not been implemented in the C# language (LINQ query syntax) - the referenced issue is still open etc. What's been done in .NET 10 is the introduction of LeftJoin and RightJoin to LINQ (no language changes).

**Comment by xiatiandegaga** [comment]
when does it( the linq of left  join) will published?.net 10 preview 5?because the left join of ef core10 don't use good,    var query = _unitOfWork.Query<PointFlow>(express)
                     .LeftJoin(
                         _unitOfWork.Query<User>(),
                         pf => pf.UserId,    
                         user => user.Id,            
                         (pf, user) => new { pf, user }
                     ).LeftJoin(
                    _unitOfWork.Query<User>(),
                    combined => combined.pf.CreatedBy,  
                    createdUser => createdUser.Id,
                    (combined, createdUser) => new { combined.pf, combined.user, createdUser }
                     ); combined is more and more，the linq of  left  join is more good than it

**Reply by xiatiandegaga** [reply]
@CyrusNajmabadi @roji

**Reply by roji** [reply]
The C# LINQ syntax support for this feature hasn't yet been implemented - when it does is TBD.

**Comment by jasekiw** [comment]
@roji @CyrusNajmabadi I saw a the [LDM notes](https://github.com/dotnet/csharplang/blob/main/meetings/2025/LDM-2025-02-12.md) that expressed optimism to get it in for C# 14 but it didn't make it. Is there discussion around getting it in for C# 15? I am aware the team is considering other syntax changes however I would be happy with just the left join getting in if I had to choose. 

Regarding breaking changes for providers I have a few ideas:

1. Generate the old `GroupJoin` expression tree equivalent so that it's compatible on day 1 with a project level toggle to turn on the new expression version or a project level compatibility version settings (  becomes more useful in further evolvement of the query syntax and expressions).

3. Provide an alternative method like `expression.V2` (I don't know if this is viable or not in terms of performance, complexity)

**Reply by roji** [reply]
There's a lot going on so I'm doubtful we'll be able to fit this in, but we'll definitely consider it.

There's no breaking change here for providers, as these are all new LINQ methods - providers that don't implement support for the new operators simply don't support them. Even if the C# syntax is added, that's still not new expression tree nodes, just ordinary method calls to new methods. So nothing needs to be done here.

**Comment by mojtabakaviani** [comment]
Hi @roji, Please consider to add left and right join clauses to C# 15 query expression syntax, We can not wait a year to add it. left and right join clauses in query expression syntax very clean when have multiple joins.
